pandas/io/json/_table_schema.py

```diff
@@ -323,9 +320,12 @@ def parse_table_schema(json, precise_float):
         for field in table["schema"]["fields"]
     }
 
-    # Cannot directly use as_type with timezone data on object; raise for now
-    if any(str(x).startswith("datetime64[ns, ") for x in dtypes.values()):
-        raise NotImplementedError('table="orient" can not yet read timezone data')
+    # tz index parsing
+    tz_info = dict()
```

Comment:
Instead of this, I think it would be better to allow `datetime64[ns, tz]` in a dict with `astype()`

Reply:
I don't really understand the idea here, can you elaborate?

Reply:
Since `df = df.astype(dtypes)` is called below, and `dtypes` is a dict like:

```
dtypes = {'col1': 'datetime64[ns, UTC]', 'col2', 'int64'}
```

It would be better for `astype` to correctly handle the timezone support instead of here.

Reply:
@mroeschke I did a re-write of the pull request in light of your comment. The impacting code was affected in the `datetimes.py` module.

Previously a `TypeError` was raised when attempting to set a timezone on an `object` dtype which was inferred to be another timezone. The change I made detects for an inferred 'UTC' timezone, which is the native format for JSON conversion, and converts it, otherwise it repeats the behaviour.

```
vals = {
    "timezones": pd.date_range(
        "2020-08-30", freq="d", periods=4, tz="Europe/London"
     )
}
df = pd.DataFrame(vals)
df = df.astype({'timezones': 'datetime64[ns, UTC]'})  # works as normal
df = df.astype({'timezones': 'object'})  # works as normal
df = df.astype({'timezones': 'datetime64[ns, Europe/Berlin]'})  # now works due to new conversion check, but previously failed due to differing tz and inferred_tz.
df = df.astype({'timezones': 'object'})  # works as normal
df = df.astype({'timezones': 'datetime64[ns, Europe/London]'})  # fails as before since inferred_tz is not 'UTC'.


```